### PR TITLE
Fix Tasks getting stuck in scheduled state

### LINKED ISSUE #19622: Tasks get stuck in "scheduled" state and starved when dags with huge amount of tasks is scheduled
### Apache Airflow version

2.0.2

### Operating System

amzn linux

### Versions of Apache Airflow Providers

_No response_

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

AIRFLOW__CORE__PARALLELISM: "128"
worker_concurrency: 32
2 Celery workers
MySQL 8.0.23 RDS as a DB backend

### What happened

* Tasks get stuck in "scheduled" state for hours, in task details it says that "All dependencies are met but the task instance is not running"
* The stuck tasks are executed eventually
* Usually, at the same time, there're DAGs with >100 tasks are running
* The big dags are limited by dag-level `concurrency` parameter to 10 tasks at a time
* Workers and pools have plenty of free slots
* If big dags are switched off - starving tasks are picked up immediately, even if tasks from the big dags are still running
* In scheduler logs, the starving task do not appear in the "tasks up for execution" list
* Number of concurrent tasks that are actually running is around 30 total on both executors (out of 64 available slots)

### What you expected to happen

As there are enough slots on the workers & pools, I expect tasks that are ready and actually *can* run to be picked up and moved to queued by scheduler

### How to reproduce

This example dag should reproduce the problem on environment with at least 20-25 available slots and core parallelism of 128. The dag that will get starved is the "tester_multi_load_3". Not each task, but on my env there were holes of up to 20 minutes between tasks execution. Guess the starvation time depends on ordering (?), as I'm not adding any weights...

<details><summary>CLICK ME</summary>
<p>

```python
import os
import time
from datetime import datetime, timedelta

from airflow import DAG
from airflow.operators.python import PythonOperator


default_args = {
    'owner': 'Tester',
    'depends_on_past': False,
    'start_date': datetime(2021, 7, 17),
    'retries': 5
}


def sleep(timer):
    if not timer:
        timer = 60
    print(f'Timer is {str(timer)}')
    time.sleep(timer)


with DAG(
        dag_id=os.path.basename(__file__).replace('.py', '') + '_1',  # name of the dag
        default_args=default_args,
        concurrency=10,
        max_active_runs=5,
        schedule_interval='@hourly',
        orientation='LR',
        tags=['testers']

) as dag1:

    for i in range(150):
        t = PythonOperator(
            task_id=f'python_{i}',
            python_callable=sleep,
            op_args=[""],
            priority_weight=-100,
        )

with DAG(
        os.path.basename(__file__).replace('.py', '') + '_2',  # name of the dag
        default_args=default_args,
        concurrency=7,
        max_active_runs=2,
        schedule_interval='@hourly',
        orientation='LR',
        tags=['testers']

) as dag2:

    for i in range(150):
        t = PythonOperator(task_id=f'python_{i}',
                           python_callable=sleep,
                           op_args=[""],
                           )


with DAG(
        os.path.basename(__file__).replace('.py', '') + '_3',  # name of the dag
        default_args=default_args,
        concurrency=1,
        max_active_runs=1,
        schedule_interval='@hourly',
        orientation='LR',
        tags=['testers']

) as dag3:

    t1 = PythonOperator(task_id=f'python', python_callable=sleep, op_args=[""])

    for i in range(10):
        t2 = PythonOperator(task_id=f'python_{i}', python_callable=sleep, op_args=[""])
        t1 >> t2
        t1 = t2
```

</p>
</details>

### Anything else

Digging around the code, I found that there's a limit on the query scheduler preforms [here](https://github.com/apache/airflow/blob/10023fdd65fa78033e7125d3d8103b63c127056e/airflow/jobs/scheduler_job.py#L928) , that comes from [here](https://github.com/apache/airflow/blob/10023fdd65fa78033e7125d3d8103b63c127056e/airflow/jobs/scheduler_job.py#L1141), and actually seems to be calculated overall from the global `parallelism` value.
So actually what happens, is that scheduler queries DB with a limit, gets back a partial list of tasks that are actually cannot be executed because of the dag-level concurrency, and gets to other tasks that are able to run only when there's a window between big dags execution. Increasing the `parallelism` to 1024 solved the issue in our case.
The `parallelism` parameter in this case is very confusing, because it should indicate tasks that can be `run concurrently`, but actually limits the scheduler's ability to move tasks from scheduled to queued... 

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by tanelk**
closes #19622

The `scheduler_job` can get stuck in a state, where it is not able to queue new tasks. It will get out of this state on its own, but the time taken depends on the runtime of current tasks - this could be several hours or even days.

If the scheduler can't queue any tasks because of different concurrency limits (per pool, dag or task), then on next iterations of the scheduler loop it will try to queue the same tasks. Meanwhile there could be some scheduled tasks with lower `priority_weight` that could be queued, but they will remain waiting.

The proposed solution is to keep track of dag and task ids, that are concurrecy limited and then repeat the query with these dags and tasks filtered out.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/jobs/scheduler_job.py
Comment (tanelk) [inline airflow/jobs/scheduler_job.py]: unrelated, but corrected a variable name, that caused some confusion

**Review by ashb — DISMISSED** [review]
This offset approach won't work well with multiple schedulers running

**Comment by vapiravfif** [comment]
Apologies if this comment won't help the discussion, but in my opinion the better approach could may be something being done around the "limit" clause - 
`        query = query.limit(max_tis)` ([origin](https://github.com/apache/airflow/blob/10023fdd65fa78033e7125d3d8103b63c127056e/airflow/jobs/scheduler_job.py#L928))
and 
`            max_tis = min(self.max_tis_per_query, self.executor.slots_available)` ([origin](https://github.com/apache/airflow/blob/10023fdd65fa78033e7125d3d8103b63c127056e/airflow/jobs/scheduler_job.py#L1141))

Because it seems that the assumption that it is a better to limit query of tasks waiting to be queued by potentially available slots might not be the best practice in some cases.

**Comment by tanelk** [comment]
> Apologies if this comment won't help the discussion, but in my opinion the better approach could may be something being done around the "limit" clause - ` query = query.limit(max_tis)` ([origin](https://github.com/apache/airflow/blob/10023fdd65fa78033e7125d3d8103b63c127056e/airflow/jobs/scheduler_job.py#L928)) and ` max_tis = min(self.max_tis_per_query, self.executor.slots_available)` ([origin](https://github.com/apache/airflow/blob/10023fdd65fa78033e7125d3d8103b63c127056e/airflow/jobs/scheduler_job.py#L1141))
> 
> Because it seems that the assumption that it is a better to limit query of tasks waiting to be queued by potentially available slots might not be the best practice in some cases.

Yes reducing the value of `max_tis` will make this situation worse, but removing the `max_tis = min(...)` does not guarantee that the issue will get solved.

As long as we are using any limit value, there is a chance that none of the selected task instancess can be queued but there are some task instances beyond the limit that could have been scheduled.

I can think of 3 possible solutions, that will never cause the scheduler to get stuck:
1. Remove the limit - could risk memory issues on a large airflow installation (many dags and tasks)
2. Build all the concurrency filters into the SQL query - the task instance limit seems to be impossible/very difficult in the current data model.
3. Some sort of iterative approach to look at the task instances further down - this is proposed in this PR.

**Comment by tanelk** [comment]
> This offset approach won't work well with multiple schedulers running

@ashb I now reworked the method

File: airflow/jobs/scheduler_job.py
Comment (ashb) [inline airflow/jobs/scheduler_job.py]: This one probably feels more appropriate as a debug log

File: tests/jobs/test_scheduler_job.py
Comment (ashb) [inline tests/jobs/test_scheduler_job.py]:
```suggestion
        tis = sorted(dr1.task_instances + dr2.task_instances, key=lambda ti: ti.key)
```

No need for the comment really

**Comment by ashb** [comment]
A lot more work this way, but I think this looks good. I'll need to take a look at this again with fresher eyes next week.

How much/where have you tested this change?

**Comment by kaxil** [comment]
Just rebased on latest main branch to fix merge conflicts

**Comment by kaxil** [comment]
Tests are failing, can you take a look plz @tanelk

**Comment by tanelk** [comment]
> Tests are failing, can you take a look plz @tanelk

I will be afk for about a week. Will take a look after that.

**Comment by tanelk** [comment]
> Tests are failing, can you take a look plz @tanelk

I fixed the merge issue. The `test_scheduler_job` suite passes now.

**Comment by ephraimbuddy** [comment]
I think I prefer this to #20391 which didn't deal with pools

**Comment by ashb** [comment]
I think we need to benchmark this approach and #20391 and see which one performs better under load

**Comment by jedcunningham** [comment]
Okay, I benchmarked 2.2.3, #19747 on latest main, and #20391 on latest main with:

- 100 DAGs, 10 runs each, 10 task per DAG for 10k total TIs
- 2 workers (1 CPU core each)
- 1 scheduler (1 CPU core)

It took each this long to get through them all:

* 2.2.3: 51 minutes
* 19747: 38 minutes
* 20391: 38 minutes

#19747 however does spew this into the scheduler logs ~16x per second (quite chatty!):

```
[2022-02-02 23:21:32,947] {scheduler_job.py:522} INFO - Setting the following tasks to queued state:
```

File: airflow/jobs/scheduler_job.py
Comment (tanelk) [inline airflow/jobs/scheduler_job.py:521]: Moved the noisy logging into the if block

**Comment by tanelk** [comment]
@kaxil , @ashb , @ephraimbuddy - any chance to get another review for this?

File: airflow/jobs/scheduler_job.py
Comment (ashb) [inline airflow/jobs/scheduler_job.py:282]: Hmmm, I wonder if this needs to include map_index now so,

```suggestion
        task_concurrency_map: DefaultDict[Tuple[str, str, int], int]
```

Making the tuple (dag_id, task_id, map_index).

See https://github.com/apache/airflow/pull/21210 for a bit more info.

Comment (ashb) [reply airflow/jobs/scheduler_job.py:282]: (I don't know if it does -- just needs thinking about)

Comment (tanelk) [reply airflow/jobs/scheduler_job.py:282]: I'm not very familiar with thw new map API, so I might be off.

If you map a task over 10 values, and that task has concurrency limit of 5, then am I right, that we should run max 5 instances at a time. Based on that we should count different map_index values just as another task instances - as it is currently.

Comment (ashb) [reply airflow/jobs/scheduler_job.py:282]: >If you map a task over 10 values, and that task has concurrency limit of 5, then am I right, that we should run max 5 instances at a time. 

Yes, that is correct. each mapped task instance is a separate TI

> Based on that we should count different map_index values just as another task instances - as it is currently.

Cool!

**Review by shihabcsedu09 — APPROVED** [review]

**Comment by ephraimbuddy** [comment]
@tanelk please rebase. I guess the failing tests have been resolved in main

**Review by jedcunningham — APPROVED** [review]
We definitely need another committer to take a look at this as well.

**Review by ephraimbuddy — APPROVED** [review]

**Comment by jedcunningham** [comment]
Thanks @tanelk! Congrats on your first commit 🎉

**Pull request merged**
